### PR TITLE
feat: Add fixer for missing unicode flag in no-misleading-character-class rule

### PR DESCRIPTION
**Pull request by mathiasvr**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[x] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)
Added autofixing for missing unicode "u" flag in case regex contains surrogate pair characters.

### CODE REVIEW

**Comment by nzakas** [comment]
Thanks for the PR. I’m not sure this fix is safe. I’m general, we don’t do fixes that change how code works because it can be difficult for end users to identify if it creates a problem. In this case, I fear that the change may fall into that category.

@eslint/eslint-tsc what do others think?

**Comment by btmills** [comment]
I agree. People often run autofix on save or as a commit hook, so we want to be extra careful about fixers. This seems like a good fit for a suggested fix.

**Comment by mathiasvr** [comment]
Auto fixing code as a commit hook sounds like a terrible idea since you generally want your commits to be in the state you committed them.

**Comment by nzakas** [comment]
Good point @btmills. @mathiasvr can you change this to a suggestion?

**Comment by mathiasvr** [comment]
@nzakas Do I just change the type to `suggestion`?

https://github.com/eslint/eslint/blob/36396bad0b13c864e4ba1f4db212f03ca6622856/lib/rules/no-misleading-character-class.js#L103

**Comment by nzakas** [comment]
Exactly, this: https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions

**Comment by mathiasvr** [comment]
Okay I changed it to a suggestion. Just wondering, why would you enable this rule and not want an auto fix but only a suggestion?

**Review by btmills — CHANGES_REQUESTED** [review]
Implementation looks good! Can you add tests that verify the suggestion output is correct? For example:

```js
{
    code: "var r = /[👍]/",
    errors: [{
        messageId: "surrogatePairWithoutUFlag",
        suggestions: [{ messageId: "suggestUnicodeFlag", output: "var r = /[👍]/u" }]
    }]
},
```

> why would you enable this rule and not want an auto fix but only a suggestion?

For cases like this that would alter behavior, we want to bring them to the developer's attention as a suggestion instead of invisibly making the change in autofix, even if we're pretty sure the change is the one they want.

**Review by btmills — APPROVED** [review]
Looks good! I like how you also added tests where there was already a non-`u` flag to make sure adding the Unicode flag didn’t clobber the existing flag(s).

**Review by nzakas — APPROVED** [review]
LGTM. Thanks!

**Review by mdjermanovic — COMMENTED** [review]
I have two concerns about this change.

First, adding the `u` flag can produce syntactically invalid regular expressions. For example, this is valid without the flag:

```js
var r = /[👍]\a/;
```

When user applies the suggestion, it will cause a parsing error because `\a` isn't allowed with the `u` flag:

```js
var r = /[👍]\a/u; // Parsing error: Invalid regular expression: /[👍]\a/: Invalid escape
```

Parsing errors will be noticeable right away, but in the case of RegExp constructors, users might become aware of the errors caused by suggestions from this rule only when they run the code.

```js
var r = new RegExp("[👍]\\a", "u"); // no parsing errors, but this will throw in runtime
```

Second, suggestions are allowed to change the behavior of the code, but that should generally be limited to the reported problem. Consider the following example:

```js
// K is U+212A

var regex = /^(?:[👍]|\W!)$/i;

console.log(
    regex.test("👍") // false
);

console.log(
    regex.test("K!") // true
);
```

Adding the `u` flag fixes the problem with `[👍]`, but since the flag applies to the whole regular expression, it also changes the behavior of `\W`, which isn't related to the reported problem:

```js
// K is U+212A

var regex = /^(?:[👍]|\W!)$/iu;

console.log(
    regex.test("👍") // true
);

console.log(
    regex.test("K!") // false!
);
```

For that reason, I think that suggestions that add the `u` flag would be more suitable for the `require-unicode-regexp` rule, as proposed in https://github.com/eslint/eslint/issues/15089.

**Comment by mathiasvr** [comment]
@mdjermanovic I don't see a problem with allowing a suggestion to add the flag as changes might be necessary in either case. If wanting to allow this behaviour the rule should not be disabled. If not having suggestion for this rule, I think the require-unicode-regexp rule should also not have it since it will require manual fixes much more often.

**Comment by mdjermanovic** [comment]
> I don't see a problem with allowing a suggestion to add the flag as changes might be necessary in either case. If wanting to allow this behaviour the rule should not be enabled.

My concern about this suggestion is its scope, because it adds a flag that applies to the whole regular expression, not just to the reported character class. The flag correctly fixes a specific problem reported by this rule, but it can also have side effects on the rest of the pattern, so it could introduce new problems that the user may not be aware of.

>  I think the require-unicode-regexp rule should also not have it since it will require manual fixes much more often.

That's a good point, maybe we shouldn't provide suggestions in that rule either, except when we're sure that the suggested fix doesn't require any further manual fixes (i.e., the behavior with and without the flag is same).

**Comment by nzakas** [comment]
> My concern about this suggestion is its scope, because it adds a flag that applies to the whole regular expression, not just to the reported character class. The flag correctly fixes a specific problem reported by this rule, but it can also have side effects on the rest of the pattern, so it could introduce new problems that the user may not be aware of.

I’m not sure this is significant enough to change the path forward here. Suggestions were created specifically for situations where the fixes will potentially cause a change in behavior. In this case, we can’t limit the scope of the change, but it seems like the suggestion is still valid.

**Comment by mdjermanovic** [comment]
> Suggestions were created specifically for situations where the fixes will potentially cause a change in behavior. In this case, we can’t limit the scope of the change, but it seems like the suggestion is still valid.

This fix is way beyond the scope of the reported problem. The suggestion message should at least indicate that the fix may have side effects on behavior that is unrelated to the reported problem, and also that it can affect the validity of the regular expression (assuming that we're not checking if the regex will be valid with the `u` flag, as currently implemented).

**Comment by nzakas** [comment]
It seems like we aren’t quite connecting on this. All suggestions have possible side effects, so it doesn’t make sense to mention that in a suggestion.

It sounds like you might be saying that the fix could create an invalid regular expression? Can you give an example of that so we can discuss?

@btmills you also approved this PR. What do you think?

**Comment by mdjermanovic** [comment]
> It seems like we aren’t quite connecting on this. All suggestions have possible side effects, so it doesn’t make sense to mention that in a suggestion.

By side effects, I mean changes not related to the reported problem.

Consider this example, user wants a regex that matches `"👍"` or any string with `length=3`.


> It sounds like you might be saying that the fix could create an invalid regular expression? Can you give an example of that so we can discuss?

**Comment by mdjermanovic** [comment]
> It seems like we aren’t quite connecting on this. All suggestions have possible side effects, so it doesn’t make sense to mention that in a suggestion.

By side effects, I mean changes in behavior not related to the reported problem. I wouldn't expect that from a suggestion.

Consider this example, user wants a regex that matches `"👍"` or any string with `length=3`.

```js
const regex = /^(?:[👍]|.{3})$/;

regex.test("👍"); // false. This is a bug, the problem is [👍]

regex.test("abc"); // true

regex.test("👶👶👶"); // false
```

After the suggestion is applied:

```js
const regex = /^(?:[👍]|.{3})$/u;

regex.test("👍"); // true, the bug is fixed

regex.test("abc"); // true

regex.test("👶👶👶"); // true! This is a side effect, .{3} now works differently
```



> It sounds like you might be saying that the fix could create an invalid regular expression? Can you give an example of that so we can discuss?

I left one example in https://github.com/eslint/eslint/pull/15279#pullrequestreview-821417600, here's a similar one:

```js
/[👍]\R/; // syntactically valid

/[👍]\R/u; // syntax error
```

Additionally, the second version could become valid in the future ([proposal-regexp-r-escape](https://github.com/tc39/proposal-regexp-r-escape)), but with completely different semantics than the original one.

**Comment by mathiasvr** [comment]
@mdjermanovic I think this is why we now do it as a suggestion and not an automatic fix so that the side effects doesn't happen without explicitly it.

If a user enable this rule they should also want all the side effects as well, since they will ultimately have to disable it if not, no matter if it is auto-fixable, suggestion or just a warning.

**Comment by mdjermanovic** [comment]
> If a user enable this rule they should also want all the side effects as well, since they will ultimately have to disable it if not, no matter if it is auto-fixable, suggestion or just a warning.

Not necessarily. There are other ways to fix the problem, for example `[👍abc]` -> `(?:👍|[abc])`.

> I think this is why we now do it as a suggestion and not an automatic fix, so the side effects doesn't happen without explicitly applying it

Yes, but there's no notice about side effects. In the `/^(?:[👍]|.{3})$/` example, the reported problem is that `[👍]` doesn't match with `"👍"`.  After the suggestion is applied, matching `"👍"`  is the expected change in the behavior (not a side effect). All other changes, like matching 6 string characters with `.{3}`, may not be expected and the user may not be aware of them (those are side effects).

**Comment by mathiasvr** [comment]
The problem of not matching the emoji cannot be fixed without the u flag right? We agree that the suggestion of adding that flag may not be enough to fix the whole regex, but how will you avoid a lint error without disabling the rule or adding the flag?
I mean if the user wants the rule enabled they have to fix the regex somehow right?

**Comment by nzakas** [comment]
I don’t see the introduction of side effects as a problem here. As already discussed, suggestions are allowed to have side effects and it’s up to the end user to verify that their code still works. Autofixes cannot have side effects. 

If there is a more narrow fix to suggest, we should also suggest that. Part of the benefit of suggestions is that we can specify multiple options.

We definitely shouldn’t suggest something that is syntactically invalid. I was under the impression that the earlier comment had already been addressed. If that’s not the case, we should fix that.

**Comment by nzakas** [comment]
**TSC Summary:** This PR seeks to add suggestions to no-misleading-character-class, specifically adding the “u” flag if it would result in a valid regular expression. Both nzakas and btmills have approved the PR, mdjermanovic has concerns about the side effects of the suggestions. We do allow side effects in suggestions in core rules. 

**TSC Question:** Do we want to change our expectations around suggestions? (https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions) Can we merge this?

**Comment by mdjermanovic** [comment]
> The problem of not matching the emoji cannot be fixed without the u flag right?

It can be fixed by moving the character out of the class.

```js
const regex = /^(?:👍|.{3})$/;

regex.test("👍"); // true, the bug is fixed

regex.test("abc"); // true

regex.test("👶👶👶"); // still false
```

This fix changes behavior related to the reported problem, but without side effects on things that are not considered problems by this rule.

There is a similar plugin rule `regexp/no-misleading-unicode-character` and its opt-in autofix works this way, here are [test cases](https://github.com/ota-meshi/eslint-plugin-regexp/blob/8c01e8669d40d502d8ba60548358331781cb0d72/tests/lib/rules/no-misleading-unicode-character.ts#L51-L98).

**Comment by mdjermanovic** [comment]
Per 16-December-2021 TSC Meeting, this PR is accepted. The conclusion is that suggestions are allowed to change the behavior of the code even when the said behavior isn't directly related to the reported problem, and that there is no need to add a note about possible side effects in the suggestion message. On the other hand, suggestions that produce syntactically invalid code should be avoided.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
Per the previous comment, we should check if the regex would be syntactically valid with the `u` flag.

For example, rule shouldn't provide the `u` flag suggestion for the following regex literal:

```js
/^[👍]\a$/;
```

For that purpose, we can use `regexpp.RegExpValidator`. Here's [an example](https://github.com/eslint/eslint/blob/775d18138244a28ebe1cb92849cd0f4e8cd27672/lib/rules/prefer-regex-literals.js#L327) in `prefer-regex-literals` rule.

File: lib/rules/no-misleading-character-class.js [hunk 1]
```diff
@@ -190,7 +201,19 @@ module.exports = {
                     const flags = getStringIfConstant(flagsNode, scope);
 
                     if (typeof pattern === "string") {
-                        verify(node, pattern, flags || "");
+                        verify(node, pattern, flags || "", fixer => {
+                            if (node.arguments.length === 1) {
+                                return fixer.insertTextAfterRange(patternNode.range, ', "u"');
+                            }
```
Comment (mdjermanovic) [inline lib/rules/no-misleading-character-class.js:207]: First argument may be parenthesised:

```js
new RegExp(("[👍]")); // regex is /[👍]/
```

In that case, the suggested fix wouldn't work as intended:

```js
new RegExp(("[👍]", "u")); // regex is /u/
```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by mdjermanovic** [comment]
> If there is a more narrow fix to suggest, we should also suggest that. Part of the benefit of suggestions is that we can specify multiple options.

This was also discussed in the TSC meeting, and we agreed that it isn't necessary for this PR. It's fine to implement only the `u` flag suggestion for the start, and (maybe) consider adding other suggestions later.

**Comment by nzakas** [comment]
@mathiasvr are you still working on this?

**Comment by mathiasvr** [comment]
@nzakas Sorry, I don't have time to look into finishing the requested changes at the moment.
